File: Introduction_to_OpenMP/Solutions/ftn/src/pi-omp-v2.f90
! Calculating Pi with using method: 
!     integral 4.0/(1+x^2) dx = p	     
! 	   OpenMP Version 2	     
program pi_omp_v1
implicit none
integer(kind=8) :: num_steps = 100000000000
integer(kind=8) i,tid,nthreads,omp_get_num_threads,omp_get_thread_num
real*8 :: PI_ref=3.1415926535897932,step,x,pi
real*8 :: summation=0.0
real*8   :: start,finish,omp_get_wtime
!call cpu_time(start)
step = 1/(1.0 * num_steps)
start=omp_get_wtime()
!$omp parallel do private(x) reduction(+:summation) shared(step)
	do i =1,num_steps
		x=((i-1)+0.5) * step;
		summation=summation + (4.0/(1.0+(x*x)))
	end do
!$omp end parallel do
	pi=step*summation

!call cpu_time(finish)
finish=omp_get_wtime()
	write(*,*) "Estimated value of pi =",pi
	write(*,*) "Error =", abs(PI_ref - pi)
	write(*,*) "Compute time=", (finish-start)," seconds"
end

File: Introduction_to_OpenMP/Solutions/c/src/pi-omp-v2.c
/*************************************
 * Calculating Pi with using method: *
 *  integral 4.0/(1+x^2) dx = p	     *
 * 	   Serial Version	     *
 ************************************/
#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

// Reference value to estimate error
double PI_ref=3.1415926535897932;

static long long num_steps=100000000000;
double step;
int main (int argc, char* argv[]){
	double start,stop,diff_time;
	long long i;
	double x,pi,sum=0.0;

	start= omp_get_wtime();  
	step = 1.0 / (double) num_steps;
#pragma omp parallel for private(x) reduction(+:sum)        
	//Work will be partitioned uniformly with the last 
	//thread getting the extra (in case the remainder(num_steps,nthreads) != 0 	
	//note: i is scoped as private by default as its the index of the for loop immidiately after pragma
	for (i=0; i<num_steps; i++){
		x=(i+0.5) * step;
		sum = sum + 4.0/(1.0+(x*x));
	}
	//Master finally evaluates pi
	pi = step * sum ;
	stop= omp_get_wtime();

	diff_time = (double) (stop-start);
	printf ("Estimated value of pi = %2.15lf\n",pi);
	printf ("Error = %2.15lf\n",fabs(PI_ref - pi));
	printf ("Compute time= %2.5lf seconds\n",diff_time);

}
